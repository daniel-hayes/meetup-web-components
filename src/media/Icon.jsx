import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import { MEDIA_SIZES } from '../utils/designConstants';
import { VALID_SHAPES } from 'swarm-icons/dist/js/shapeConstants';

export const ICON_CLASS = 'svg';
export const SVG_THIN_STYLE = '--small';
export const ICON_CIRCLED_CLASS = 'svg--circled';

const SMALL_ICON_VARIANT_WHITELIST = VALID_SHAPES.filter(
	s => !s.startsWith('external') && !s.startsWith('meetup') // no third party icons // logos use same path for `xs`
);

/**
 * @param {String} shape - icon shape
 * @param {String} size - icon size
 * @returns {String} icon name (with or without suffix)
 */
export const getIconShape = (shape, size) => {
	if (!SMALL_ICON_VARIANT_WHITELIST.includes(shape)) {
		return shape;
	}

	const suffix = size === 'xxs' || size === 'xs' || size === 's' ? SVG_THIN_STYLE : '';
	return `${shape}${suffix}`;
};

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
class Icon extends React.PureComponent {
	render() {
		const { className, shape, size, color, style, circled, ...other } = this.props;

		const classNames = cx(
			ICON_CLASS,
			`svg--${shape}`,
			{ [ICON_CIRCLED_CLASS]: circled },
			className
		);

		const sizeVal = MEDIA_SIZES[size];

		const allStyles = style || {};
		if (color) {
			allStyles.fill = color;
		}

		return (
			<span className={classNames}>
				<svg
					preserveAspectRatio="xMinYMin meet"
					width={sizeVal}
					height={sizeVal}
					viewBox={`0 0 ${sizeVal} ${sizeVal}`}
					className="svg-icon valign--middle"
					role="img"
					style={allStyles}
					{...other}
				>
					<use xlinkHref={`#icon-${getIconShape(shape, size)}`} />
				</svg>
			</span>
		);
	}
}

Icon.defaultProps = {
	size: 's',
};

Icon.propTypes = {
	/** The name of the icon shape to render */
	shape: PropTypes.oneOf(VALID_SHAPES).isRequired,

	/** Which of our media sizes to render the icon at */
	size: PropTypes.oneOf(Object.keys(MEDIA_SIZES)).isRequired,

	/** Whether the icon is outlined with a circle */
	circled: PropTypes.bool,

	/** What color the icon should be filled with */
	color: PropTypes.string,
};

export default Icon;
