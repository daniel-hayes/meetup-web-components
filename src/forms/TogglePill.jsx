import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import Icon from '../media/Icon';
import ToggleWrapper from '../utils/components/ToggleWrapper';

export const TOGGLE_PILL_CLASS = 'toggleButton';

/**
 * Toggle Pill component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_toggle-pill.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#togglePills}
 * @module TogglePill
 */
export class TogglePill extends React.PureComponent {
	render() {
		const {
			isActive,
			children,
			className,
			labelClassName,
			topic,
			id,
			small,
			name,
			useRadio,
			value,
			onChange,
			...other
		} = this.props;

		delete other.large;

		const inputType = useRadio ? 'radio' : 'checkbox';

		const classNames = cx(
			TOGGLE_PILL_CLASS,
			{
				'toggleButton--topic': topic,
				'toggleButton--radio': useRadio,
			},
			className
		);

		const labelClassNames = cx(
			'toggleButton-label',
			{
				'toggleButton-label--small': small,
			},
			labelClassName
		);

		const topicChildren = isActive => (
			<Icon
				className={cx('toggleButton-icon', {
					'toggleButton-icon--active': isActive,
					'toggleButton-icon--inactive': !isActive,
				})}
				shape={isActive ? 'heart' : 'heart-outline'}
				size="xxs"
				label="Active Topic Icon"
			/>
		);
		// ---

		return (
			<ToggleWrapper type={inputType} isActive={isActive} onToggle={onChange}>
				{({ tabIndex, isActive, toggleActive, onKeyUp }) => (
					<div className={classNames}>
						<input
							className="toggleButton-input visibility--a11yHide"
							type={inputType}
							id={id}
							name={name}
							value={value}
							checked={isActive}
							onChange={toggleActive}
							onKeyUp={onKeyUp}
							tabIndex={tabIndex}
							{...other}
						/>
						<label className={labelClassNames} htmlFor={id}>
							{children}
							{topic ? topicChildren(isActive) : null}
						</label>
					</div>
				)}
			</ToggleWrapper>
		);
	}
}

TogglePill.propTypes = {
	/** Adds an `id` attribute to the input, and associates it with the `<label />` */
	id: PropTypes.string.isRequired,

	/** The `name` attribute for the input */
	name: PropTypes.string.isRequired,

	/** The `value` attribute for the input */
	value: PropTypes.string.isRequired,

	children: PropTypes.node.isRequired,

	/** Whether the pill is toggled on/selected */
	isActive: PropTypes.bool,

	/** Whether to use input[type="radio"] to limit the selection to one toggle pill in a group */
	useRadio: PropTypes.bool,

	/** Whether the pill should be rendered in a way that visually indicates it's a topic */
	topic: PropTypes.bool,

	/** The class name/s to add to the `<label />` element */
	labelClassName: PropTypes.string,
};

TogglePill.defaultProps = {
	isActive: false,
};

export default TogglePill;
