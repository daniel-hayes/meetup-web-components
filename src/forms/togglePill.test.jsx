import React from 'react';
import { mount } from 'enzyme';
import TogglePill, { TOGGLE_PILL_CLASS } from './TogglePill';
import Icon from '../media/Icon';

describe('TogglePill', () => {
	const id = 'hikingCategory',
		name = 'meetupCategories',
		label = 'Hiking!',
		value = 'hiking',
		labelClass = 'someClass';

	let togglePillComponent;
	const onChangeMock = jest.fn();

	beforeEach(() => {
		togglePillComponent = mount(
			<TogglePill
				onChange={onChangeMock}
				id={id}
				name={name}
				labelClassName={labelClass}
				value={value}
			>
				{label}
			</TogglePill>
		);
	});

	afterEach(() => {
		togglePillComponent = null;
	});

	it('renders a component with expected attributes', () => {
		expect(togglePillComponent).toMatchSnapshot();
	});

	it('has appropriate toggle pill class', () => {
		expect(togglePillComponent.find(TOGGLE_PILL_CLASS)).not.toBeNull();
	});

	it('executes onChange when clicked', () => {
		const toggleInput = togglePillComponent.find('input');

		expect(onChangeMock).not.toHaveBeenCalled();
		toggleInput.simulate('change');
		expect(onChangeMock).toHaveBeenCalled();
	});

	describe('TogglePill with topic icon', () => {
		const id = 'parentingTopic',
			name = 'topics',
			label = 'Moms and Dads',
			value = 'parenting';

		beforeEach(() => {
			togglePillComponent = mount(
				<TogglePill topic id={id} name={name} value={value}>
					{label}
				</TogglePill>
			);
		});

		afterEach(() => {
			togglePillComponent = null;
		});

		it('creates an Icon component', function() {
			expect(togglePillComponent.find(Icon).length).toBeGreaterThan(0);
		});
	});

	describe('TogglePill with radio input', () => {
		const id = 'parentingTopic',
			name = 'topics',
			label = 'Moms and Dads',
			value = 'parenting';

		beforeEach(() => {
			togglePillComponent = mount(
				<TogglePill useRadio id={id} name={name} value={value}>
					{label}
				</TogglePill>
			);
		});

		afterEach(() => {
			togglePillComponent = null;
		});

		it('creates a Toggle Pill with a radio input', function() {
			expect(
				togglePillComponent.find('input[type="radio"]').length
			).toBeGreaterThan(0);
		});
	});
});
