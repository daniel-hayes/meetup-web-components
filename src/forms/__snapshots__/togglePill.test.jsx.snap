// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogglePill renders a component with expected attributes 1`] = `
<TogglePill
  id="hikingCategory"
  isActive={false}
  labelClassName="someClass"
  name="meetupCategories"
  onChange={[MockFunction]}
  value="hiking"
>
  <ToggleWrapper
    isActive={false}
    onToggle={[MockFunction]}
    type="checkbox"
  >
    <ConditionalWrap
      condition={false}
      wrap={[Function]}
    >
      <div
        className="toggleButton"
      >
        <input
          checked={false}
          className="toggleButton-input visibility--a11yHide"
          id="hikingCategory"
          name="meetupCategories"
          onChange={[Function]}
          onKeyUp={[Function]}
          tabIndex="0"
          type="checkbox"
          value="hiking"
        />
        <label
          className="toggleButton-label someClass"
          htmlFor="hikingCategory"
        >
          Hiking!
        </label>
      </div>
    </ConditionalWrap>
  </ToggleWrapper>
</TogglePill>
`;
